src/lxml/sax.py

```diff
@@ -200,6 +200,7 @@ def _recursive_saxify(self, element, prefixes):
                 content_handler.characters(element.tail)
             return
 
+        element_prefixes = dict((v, k) for (k, v) in element.nsmap.items())
```

Comment:
One problem here: if there is more than one prefix defined for a namespace URI, it becomes non-deterministic which one gets used. This is not as unusual as it might seem. For example, it's not uncommon to have a namespace as both the default and a prefixed namespace, e.g. when namespaced attributes are involved.

Another thing is that the nsmap is not cached internally but gets freshly created on each request. That's not hugely expensive (and we're already copying an entire tree through Python here anyway), but it would still add up since it's done for every element. Not sure if there is anything easy that could be done about it here.

And finally, while looking through the original code, I noticed something that looks like a bug. The `prefixes` mapping is collected globally, whereas it can actually differ between separate subtrees. If one subtree overrides a prefix that is defined further up, the current `ElementTreeProducer` could ignore that and start returning an incorrect prefix. I think the namespace handling should take advantage of the recursion to keep track of the currently valid prefix mapping. It has to do that anyway since it must report namespace SAX events.

Would you like to look into that problem? It might actually lead to an even better implementation.

Reply:
Sorry this took such long time, I ended up rewriting xmldiff from scratch, and not using the saxifier...

Anyway, I pushed another, better, attempt that should deal with the recursion. It also should prefer to not use a prefix for a default namespace. I don't have tests for that yet, will work on it shortly.

Reply:
And I added a test, and now make the saxify prefer the prefix that the element actually had.